rust/core-lib/src/watcher.rs

```diff
@@ -12,108 +12,501 @@
+            (self.recursive || &self.path == path) ||
+                path.parent() == Some(&self.path)
+        } else {
+            false
+        };
+
+        if let Some(ref filter) = self.filter {
+            general_case && filter(path)
+        } else {
+            general_case
+        }
+    }
+}
+
+impl Notify for RpcPeer {
+    fn notify(&self) {
+        self.schedule_idle(::tabs::WATCH_IDLE_TOKEN);
     }
+}
 
-    //TODO impl unwatch, when we add in watching of opened files
+impl fmt::Debug for Watchee {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Watchee path: {:?}, r {}, t {} f {}",
+               self.path, self.recursive, self.token.0, self.filter.is_some())
+    }
+}
+
+fn mode_from_bool(b: bool) -> RecursiveMode {
+    if b {
+        RecursiveMode::Recursive
+    } else {
+        RecursiveMode::NonRecursive
+    }
 }
 
-/// Checks the predicate against the various event cases
-fn apply_filter<F>(filter: &F, event: &DebouncedEvent) -> bool
-    where F: Fn(&Path) -> bool,
-{
+// Debounced event does not implement clone
+// TODO: remove if https://github.com/passcod/notify/pull/133 is merged
+fn clone_event(event: &DebouncedEvent) -> DebouncedEvent {
     use self::DebouncedEvent::*;
+    use notify::Error::*;
     match *event {
-        NoticeWrite(ref p) | NoticeRemove(ref p) | Create(ref p) |
-            Write(ref p) | Chmod(ref p) | Remove(ref p) => {
-                filter(p)
+        NoticeWrite(ref p) => NoticeWrite(p.to_owned()),
+        NoticeRemove(ref p) => NoticeRemove(p.to_owned()),
+        Create(ref p) => Create(p.to_owned()),
+        Write(ref p) => Write(p.to_owned()),
+        Chmod(ref p) => Chmod(p.to_owned()),
+        Remove(ref p) => Remove(p.to_owned()),
+        Rename(ref p1, ref p2) => Rename(p1.to_owned(), p2.to_owned()),
+        Rescan => Rescan,
+        Error(ref e, ref opt_p) => {
+            let error = match *e {
+                PathNotFound => PathNotFound,
+                WatchNotFound => WatchNotFound,
+                Generic(ref s) => Generic(s.to_owned()),
+                Io(ref e) => Generic(format!("{:?}", e)),
+            };
+            Error(error, opt_p.clone())
+        }
+    }
+}
+
+#[cfg(test)]
+extern crate tempdir;
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::ffi::OsStr;
+    use std::thread;
+    use std::fs;
+    use std::sync::mpsc;
+    use std::time::{Duration, Instant};
+    use std::io::Write;
+
+    impl PartialEq<usize> for WatchToken {
+        fn eq(&self, other: &usize) -> bool {
+            self.0 == *other
+        }
+    }
+
+    impl From<usize> for WatchToken {
+        fn from(err: usize) -> WatchToken {
+            WatchToken(err)
+        }
+    }
+
+    impl Notify for mpsc::Sender<bool> {
+        fn notify(&self) {
+            self.send(true).expect("send shouldn't fail")
+        }
+    }
+
+    // Sleep for `duration` in milliseconds
+    pub fn sleep(duration: u64) {
```

Comment:
I would say `millis` rather than `duration` as a more precise characterization of units.